File: 31-next-permutation/next-permutation.java
class Solution {
    public void nextPermutation(int[] nums) {
        List<Integer> nu=new ArrayList<>();

        int n=nums.length;
        int b=-1;
        for(int i=n-1;i>0;i--){
            if(nums[i]>nums[i-1]){
                System.out.println(i-1);
                b=i-1;
                break;
            }
        }

        if(b==-1){
            Arrays.sort(nums);
            return ;
        }

        for(int i=n-1;i>b;i--){
            if(nums[i]>nums[b]){
                System.out.println(i);
                int temp=nums[b];
                nums[b]=nums[i];
                nums[i]=temp;
                break;
            }
        }

        for(int i=n-1;i>b;i--){
            nu.add(nums[i]);
        }
        
        int c=b+1;
        for(int i:nu){
           nums[c++]=i;
        }

    }
}

File: 31-next-permutation/next-permutation.py
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        

        b = -1
        n = len(nums)
        for i in range(n - 2, -1, -1):
            if nums[i] < nums[i + 1]:
                b = i
                break
    
        if b == -1:
            return nums.reverse()
        
        maxi = 10 ** 10
        maxiid = -1
        for i in range(b + 1, n):
            if nums[i] > nums[b] and nums[i] < maxi:
                maxi = nums[i]
                maxiid = i
        
        nums[b], nums[maxiid] = nums[maxiid], nums[b]
        temp = nums[b + 1:]
        temp.sort()
        i = b + 1
        for no in temp:
            nums[i] = no
            i += 1
        return nums